llvm/lib/Transforms/Utils/SYCLModuleSplit.cpp

```diff
@@ -0,0 +1,935 @@
+      GVs, MD.getModule(), ModuleEntryPoints, CG, IncludeFunctionPredicate);
+
+  ModuleDesc SplitM = extractSubModule(MD, GVs, std::move(ModuleEntryPoints));
+  LLVM_DEBUG(SplitM.dump(dbgs()));
+  SplitM.cleanup();
+
+  return SplitM;
+}
+
+class ModuleCopier : public ModuleSplitterBase {
+public:
+  using ModuleSplitterBase::ModuleSplitterBase; // to inherit base constructors
+
+  ModuleDesc nextSplit() override {
+    ModuleDesc Desc{releaseInputModule(), nextGroup()};
+    // Do some basic optimization like unused symbol removal
+    // even if there was no split.
+    Desc.cleanup();
+    return Desc;
+  }
+};
+
+class ModuleSplitter : public ModuleSplitterBase {
+public:
+  ModuleSplitter(ModuleDesc MD, EntryPointGroupVec GroupVec)
+      : ModuleSplitterBase(std::move(MD), std::move(GroupVec)),
+        CG(Input.getModule()) {}
+
+  ModuleDesc nextSplit() override {
+    return extractCallGraph(Input, nextGroup(), CG);
+  }
+
+private:
+  DependencyGraph CG;
+};
+
+} // namespace
+
+namespace llvm {
+
+std::optional<IRSplitMode> convertStringToSplitMode(StringRef S) {
+  static const StringMap<IRSplitMode> Values = {{"kernel", SPLIT_PER_KERNEL},
+                                                {"source", SPLIT_PER_TU},
+                                                {"auto", SPLIT_AUTO},
+                                                {"none", SPLIT_NONE}};
+
+  auto It = Values.find(S);
+  if (It == Values.end())
+    return std::nullopt;
+
+  return It->second;
+}
+
+void dumpEntryPoints(raw_ostream &OS, const EntryPointSet &C,
+                     std::string_view Msg) {
+  constexpr size_t INDENT = 4;
+  OS.indent(INDENT) << "ENTRY POINTS"
+                    << " " << Msg << " {\n";
+  for (const Function *F : C)
+    OS.indent(INDENT) << "  " << F->getName() << "\n";
+
+  OS.indent(INDENT) << "}\n";
+}
+
+void dumpEntryPoints(raw_ostream &OS, const Module &M,
+                     bool OnlyKernelsAreEntryPoints, std::string_view Msg) {
+  constexpr size_t INDENT = 4;
+  OS.indent(INDENT) << "ENTRY POINTS (Module)"
+                    << " " << Msg << " {\n";
+  for (const auto &F : M)
+    if (isEntryPoint(F, OnlyKernelsAreEntryPoints))
+      OS.indent(INDENT) << "  " << F.getName() << "\n";
+
+  OS.indent(INDENT) << "}\n";
+}
+
+void ModuleDesc::cleanup() {
+  // Externalize them so they are not dropped by GlobalDCE
+  for (Function &F : *M)
+    if (F.hasFnAttribute("indirectly-callable"))
+      F.setLinkage(GlobalValue::LinkageTypes::ExternalLinkage);
+
+  ModuleAnalysisManager MAM;
+  MAM.registerPass([&] { return PassInstrumentationAnalysis(); });
+  ModulePassManager MPM;
+  // Do cleanup.
+  MPM.addPass(GlobalDCEPass());           // Delete unreachable globals.
+  MPM.addPass(StripDeadDebugInfoPass());  // Remove dead debug info.
+  MPM.addPass(StripDeadPrototypesPass()); // Remove dead func decls.
+  MPM.run(*M, MAM);
+
+  // Original module may have named metadata (spirv.ExecutionMode) referencing
```

Comment:
This comment should be moved to the function definition/declaration. It could be renamed to something "fixupNamedMetadataReferencingKernels" to better indicate the intent when its call site is separate from its documentaiton.

Reply:
Removed that function.